#!/bin/bash
set -e
INDIR="/w/00/g/g00/sangyupeng/xiangyang/combinegvcf/delect"
OUTDIR="/w/00/g/g00/sangyupeng/xiangyang/GATK_vcf"
bed_dir="/w/00/g/g00/sangyupeng/xiangyang/reference/bed"
reference="/w/00/g/g00/sangyupeng/xiangyang/reference/zsp193a-LH.LG.fasta"
JAVA="/data/apps/jdk/jdk1.8.0_131/bin"
for bed in $bed_dir/*
do
	temp=${bed##*/}
	chr=${temp%.*}
	inputfile=${temp%.*}.234.g.recode.vcf.gz
	outputfile=${temp%.*}.234.vcf.gz
	echo "cd $OUTDIR;source ~/.bashrc;$GATK/gatk --java-options \"-Xmx36g\" GenotypeGVCFs -R $reference -L $bed -V $INDIR/$inputfile -O $outputfile --heterozygosity 0.01 -all-sites --indel-heterozygosity 0.0025 --allow-old-rms-mapping-quality-annotation-data" >> genotype.sh		
done
