#!/bin/bash
set -e
reference="/w/00/g/g00/sangyupeng/xiangyang/reference/zsp193a-LH.LG.fasta"
outdir="/w/00/g/g00/sangyupeng/xiangyang/reference/true_bed"
tool_dir="/w/00/u/user201/software/seqbility-20091110"
BWA="/data/apps/bwa/bwa-0.7.17"
SAMTOOLS="/data/apps/samtools/1.9/bin"
PICARD="/data/apps/picard/2.18.11"

###The first step is to create bed file that based on the information of uniquely mapepd regions by reads created by seqbility-20091110
###Step1: split the genome into reads, with the length of each read is 100 bases

$tool_dir/splitfa $reference 100 > $outdir/zsp193a.reads

###Step2: merge the reads together and align the short reads back to the genome with bwa aln
$BWA/bwa aln -R 1000000 -O 3 -E 3 $reference $outdir/zsp193a.reads > $outdir/zsp193a.sai
$BWA/bwa samse $reference $outdir/zsp193a.sai $outdir/zsp193a.reads > $outdir/zsp193a.sam
###Step3: generate the raw mask file and then set the stringency to make the final mask files
perl /w/00/u/user201/software/seqbility-20091110/gen_raw_mask.pl $outdir/zsp193a.sam > $outdir/zsp193a.rawMask_100.fa

/w/00/u/user201/software/seqbility-20091110/gen_mask -l 100 -r 0.5 $outdir/zsp193a.rawMask_100.fa > $outdir/zsp193a.rawMask_100_0.5.fa
rm $outdir/zsp193a.sai $outdir/zsp193a.sam $outdir/zsp193a.reads
/w/00/u/user201/software/seqbility-20091110/gen_mask -l 100 -r 0.5 $outdir/zsp193a.rawMask_100_0.5.fa > $outdir/zsp193a.Mask_100_0.5.fa
###Step4: generate the chr.txt which contains two columns (chr pos)
sh 12bed_1.txt
sh 12bed_2.txt
sh 12bed_3.txt

###Step5: use several commands to transper the output from the above step to the bed file, which contains the information of the uniquely mapped positions on the specific chromosome

sh 12bed_4.txt 





 
