.libPaths("F:/360MoveData/Users/dell/Documents/R")
rm(list=ls())
library(gggenes)
library(Rmisc)
library(ggplot2)
library(raster)
library(sf)
library(dplyr)
library(forcats)
library(tidyr)
library(data.table)
library(cowplot)
library(qqman)
require(gridExtra)
library(tidyverse)
library(rnaturalearth)
library(scatterpie)
library(sf)
setwd("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/manhatton/manhatton_input")
outdir="F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/picture"
manhatton_snp=read.table("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/manhatton/manhatton_snp.txt",head=T)
manhatton_snp=subset(manhatton_snp,remain=="TRUE")
### create manhatton-out list
plot_manhatton <- list()
###########  manhatton ########### 
for (i in c(1:4,6:12,14:19)){
  infile=paste("data3_filter1_bio",i,".txt",sep='')
  gwasResults=fread(infile,head=T,sep=",")
  BIO=paste("BIO",i,sep='')

  chr_len <- gwasResults %>% 
    group_by(CHR) %>% 
    summarise(chr_len=max(BP))

  chr_pos <- chr_len  %>%
    mutate(total = cumsum(chr_len) - chr_len) %>%
    select(-chr_len)

  Snp_pos <- chr_pos %>%
    left_join(gwasResults, ., by="CHR") %>%
    arrange(CHR, BP) %>%
    mutate( BPcum = BP + total)
  colors=rep(c("#3B6EFF", "#A4C8FE"), 22 )
  if (i<=11){
    colors=rep(c("#D8151D","#FC7034"), 22 )
  }

  INDEL=subset(Snp_pos,TYPE=="INDEL")
  SNP=subset(Snp_pos,TYPE=="SNP")
  SV=subset(Snp_pos,TYPE=="SV")
  INDEL_q=subset(INDEL,Q<=0.05)
  SNP_q=subset(SNP,Q<=0.05)
  SV_q=subset(SV,Q<=0.05)
  max_p_INDEL=-log10(max(INDEL_q$P))
  max_p_SNP=-log10(max(SNP_q$P))
  max_p_SV=-log10(max(SV_q$P))	
### There is an error in BIO7, which should be handled separately
  if (i==7) {
    Notable_line=min(max_p_INDEL,max_p_SNP)
  }else{
    Notable_line=min(max_p_INDEL,max_p_SNP,max_p_SV)  
  }
  

  gene_data=subset(manhatton_snp,BIO==i)
  gene_data$P=subset(Snp_pos,Snp_pos$ID%in%gene_data$ID)$P
  gene_data$BPcum=subset(Snp_pos,Snp_pos$ID%in%gene_data$ID)$BPcum
  
  X_axis <-  Snp_pos %>% group_by(CHR) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )
  #### make plot  
  plot_manhatton [[i]] <- ggplot(Snp_pos, aes(x=BPcum, y=-log10(P))) +
    geom_point(data=SNP, aes(color=as.factor(CHR),shape=as.factor(TYPE)), alpha=0.6, size=0.7) +
    geom_point(data=INDEL, aes(color=as.factor(CHR),shape=as.factor(TYPE)), alpha=0.7, size=1.1) +
    geom_point(data=SV, aes(color=as.factor(CHR),shape=as.factor(TYPE)),  alpha=0.7,size=1.1) +
    geom_point(data=gene_data, aes(x=BPcum, y=-log10(P),shape=as.factor(TYPE)), color="black", size=1.1) +
    geom_text(data=gene_data, aes(x=BPcum, y=-log10(P)-0.5,label=GENE,fontface='italic'),vjust=0.5,hjust=0,size=4.5,family="serif",color="black",check_overlap=F)+
    ### GEOM_TEXT_REPEL 
    scale_color_manual(values = colors) +
    scale_shape_manual(values=c(17,16,15))+
    scale_x_continuous( label = c("LG01","LG02","LG03","LG04","LG05","LG06","LG07","LG08","LG09","LG10","LG11","LG12","LG13","LG14","LG15","LG16","LG17","LG18","LG19"), breaks= X_axis$center ) +
    labs(x="",y=paste("-log10(P):BIO",i,sep="")) +
    scale_y_continuous(limits=c(3,-log10(min(Snp_pos$P))+0.5),expand = c(0, 0) ) +
    geom_hline(yintercept = Notable_line, color = "grey", size = 0.5, alpha=0.8, linetype = "twodash")+
    theme_bw() +
    theme(
      text=element_text(family="serif"),
      plot.title=element_text(size = 10,vjust = -0.2,hjust=0.9,face = 'bold'),
      axis.ticks.length = unit(0.25,"lines"),axis.ticks=element_line(colour="black",unit(0.6,"line")),
      axis.text.x=element_text(vjust=1,hjust=0,size=8,family="serif",colour="black"),
      axis.text.y=element_text(vjust=1,hjust=0.9,size=8.5,family="serif",colour="black"),
      axis.title.y = element_text(size =11),
      axis.line.x=element_line(colour="black"),
      axis.line.y=element_line(colour="black"),
      panel.border=element_blank(),
      legend.position = "none",
      axis.title.x = element_text(size = 11),
      panel.background=element_rect(fill="white"),
      plot.background = element_rect(fill = "white"),
      panel.grid.major =element_blank(), panel.grid.minor = element_blank(),
      plot.margin=unit(c(0.5,0.5,0.5,0.5),"mm"))
}

###########  freq-map ########### 
setwd("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/freq")
#### read .txt
files_txt=list.files(path="F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/freq",full.names = F,pattern = "txt")
jiangyu=c("#EFF7B5","#CEECB2","#96D6B9","#5EC0C0","#2DA4C1","#2080B7","#2254A3","#134BA3")
wendu=c("#4575b7","#5588bd","#a7d6e6","#f2f8dc","#FFE3A7","#EEAF9F","#E87C5F","#D82D17")
country_shp=sf::st_read("E:/PK/15GF/3055final/world/world/world.shp")
country_shp <- st_transform(country_shp, "+init=epsg:4508")
coord=read.table("F:/360MoveData/Users/Desktop/data3/freq/coord.txt",head=F)
#### create 8 groups
creategroup <- function(tiff){
  colnames(tiff)=c("x","y","bio_value")
  var=max(tiff$bio_value)-min(tiff$bio_value)
  min=floor(min(tiff$bio_value))
  if (min>0){
    group8=paste(min+floor(var/8*7),'+',sep='')
    group7=paste(min+floor(var/8*6),'~',min+floor(var/8*7),sep='')
    group6=paste(min+floor(var/8*5),'~',min+floor(var/8*6),sep='')
    group5=paste(min+floor(var/8*4),'~',min+floor(var/8*5),sep='')
    group4=paste(min+floor(var/8*3),'~',min+floor(var/8*4),sep='')
    group3=paste(min+floor(var/8*2),'~',min+floor(var/8*3),sep='')
    group2=paste(min+floor(var/8*1),'~',min+floor(var/8*2),sep='')
    group1=paste(min,'~',min+floor(var/8*1),sep='')  
    
  }else {
    group8=paste(min+floor(var/8*7),'+',sep='')
    group7=paste(min+floor(var/8*6),'~',min+floor(var/8*7),sep='')
    group6=paste(min+floor(var/8*5),'~',min+floor(var/8*6),sep='')
    group5=paste(min+floor(var/8*4),'~',min+floor(var/8*5),sep='')
    group4=paste(min+floor(var/8*3),'~',min+floor(var/8*4),sep='')
    group3=paste(min+floor(var/8*2),'~',min+floor(var/8*3),sep='')
    group2=paste(min+floor(var/8*1),'~',min+floor(var/8*2),sep='')
    group1=paste(min,'~',min+floor(var/8*1),sep='')
  }
  tiff$levels=ifelse(tiff$bio_value>=min+floor(var/8*7),group8,
                     ifelse(tiff$bio_value>=min+floor(var/8*6),group7,
                            ifelse(tiff$bio_value>=min+floor(var/8*5),group6,
                                   ifelse(tiff$bio_value>=min+floor(var/8*4),group5,
                                          ifelse(tiff$bio_value>=min+floor(var/8*3),group4,
                                                 ifelse(tiff$bio_value>=min+floor(var/8*2),group3,
                                                        ifelse(tiff$bio_value>=min+floor(var/8*1),group2,group1
                                                        )))))))
  
  
  tiff$label_order=ifelse(tiff$levels==group8,7,
                          ifelse(tiff$levels==group7,6,
                                 ifelse(tiff$levels==group6,5,
                                        ifelse(tiff$levels==group5,4,
                                               ifelse(tiff$levels==group4,3,
                                                      ifelse(tiff$levels==group3,2,
                                                             ifelse(tiff$levels==group2,1,0
                                                             )))))))
  return(tiff)
}
files_tiff=list.files(path="E:/PK/17RONA/ABStif/current",full.names = F,pattern = "tif")
text_data=c("Annual Mean Temperature","Mean Diurnal Range","Isothermality (BIO2/BIO7)","Temperature Seasonality","Maximum Temperature of Warmest Month","Minimum Temperature of Coldest Month","Temperature Annual Range (BIO5-BIO6)","Mean Temperature of Wettest Quarter","Mean Temperature of Driest Quarter",
            "Mean Temperature of Warmest Quarter","Mean Temperature of Coldest Quarter","Annual Precipitation","Precipitation of Wettest Month",
            "Precipitation of Driest Month","Precipitation Seasonality","Precipitation of Wettest Quarter","Precipitation of Driest Quarter","Precipitation of Warmest Quarter","Precipitation of Coldest Quarter")
### create map-out list
plot_map <- list()
### create gene_name list
title_name <- list()

for (i in 1:length(files_txt)){
  a=unlist(strsplit(files_txt[i],split="_"))
  gene_function=a[2]
  CHR=a[3]
  pos=a[4]
  type=a[5]
  GENE=a[6]
  REF=a[7]
  ALT=a[8]
  temp=gsub("BIO","",a[9])
  env=gsub(".txt","",temp)
  BIO=paste("E:/PK/17RONA/ABStif/current/",env,".tif",sep="")
  n=as.numeric(env)
  text=text_data[n]
  if (n > 11) {
    colors=c(jiangyu,"#A4C8FE","#2C53BF")
    bio=paste("BIO",n," (mm)",sep="")
  }else {
    colors=c(wendu,"#E5C7C7","#BA3841")
    bio=paste("BIO",n," (â„ƒ)",sep="")
  } 
  title_name[[n]]=paste(GENE," : ",CHR,"_",pos,"_",type,"(",gene_function,")",sep="")
  out_PDF=paste("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/freq/picture/",gsub(".txt",".pdf",files_txt[i]),sep="")
  out_PNG=paste("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/freq/picture/",gsub(".txt",".png",files_txt[i]),sep="")
  resdf=stack(BIO)
  resdf<-as.data.frame(resdf,xy=TRUE)%>%drop_na() 
  resdf=creategroup(resdf)
  resdf=resdf[sort(resdf$label_order,index.return=TRUE)$ix,]
  shunxu=unique(resdf$levels)
  PIE=read.table(files_txt[i],head=F)
  pie=cbind(coord,PIE)
  colnames(pie)=c("lon","lat","REF","ALT")
  names(pie)=c("lon","lat",paste("REF:",REF,sep=""),paste("ALT:",ALT,sep=""))
  plot_map[[n]]=ggplot()+
    geom_raster(aes(x=x,y=y,fill=factor(levels)),data=resdf,stat="identity")+
    geom_sf(fill="transparent",data=country_shp)+coord_sf(xlim=c(114.8, 134),ylim=c(39, 55))+
    scale_fill_manual(values=colors,xlab(label=bio),breaks=c(shunxu[1],shunxu[2],shunxu[3],shunxu[4],shunxu[5],shunxu[6],shunxu[7],shunxu[8],colnames(pie)[3],colnames(pie)[4]))+
    labs(x="Longitude",y="Latitude")+
    scale_x_continuous(limits = c(114.8, 134))+
    scale_y_continuous(limits = c(39, 55))+
    geom_scatterpie(data=pie,aes(x=lon, y=lat), pie_scale = 1.35,
                    cols=colnames(pie)[3:4],alpha=0.8,size=0.3)+
    theme_bw()+
    theme(text=element_text(family="serif"),
          axis.ticks.length = unit(0.25,"lines"),axis.ticks=element_line(colour="black",unit(0.6,"line")),
          axis.text.x=element_text(angle=0,hjust=1,vjust=0.5,size=12,colour = "black"),
          axis.text.y=element_text(size=12,colour = "black"), 
          plot.title = element_text(
            size = 15L,
            hjust = 0
          ),
          #axis.title.y=element_blank(),
          #axis.title.x=element_blank(),
          axis.title.y = element_text(size = 12),
          axis.title.x = element_text(size = 12),
          legend.key.size=unit(0.15,'inch'),
          legend.title = element_text(size=10.5, color="black",vjust=0.5, hjust=0.5),
          legend.position = c(.121,.79),
          legend.background=element_rect(colour= "grey" ,fill= "white" ,size=0.6),
          legend.text= element_text(size=8, color="black",vjust=0.5, hjust=0.5),
          panel.background=element_rect(fill="white"),
          plot.background = element_rect(fill = "white"),
          panel.grid.minor = element_line(colour = "white",size=0.1,linetype = 4),
          plot.margin=unit(c(0.5,0.5,0.5,0.5),"mm"))
  #ggsave(plot_map[[n]],filename=out_PDF,width=7.5,height=6.5,dpi=600)
}

########### liner-regression plot ########### 
current_data=read.csv("F:/360MoveData/Users/Desktop/pk_paper/picture/7RONA/important_sites/pop_current.csv",head=T)
text_data=c("Annual Mean Temperature","Mean Diurnal Range","Isothermality (BIO2/BIO7)","Temperature Seasonality","Maximum Temperature of Warmest Month","Minimum Temperature of Coldest Month","Temperature Annual Range (BIO5-BIO6)","Mean Temperature of Wettest Quarter","Mean Temperature of Driest Quarter",
            "Mean Temperature of Warmest Quarter","Mean Temperature of Coldest Quarter","Annual Precipitation","Precipitation of Wettest Month",
            "Precipitation of Driest Month","Precipitation Seasonality","Precipitation of Wettest Quarter","Precipitation of Driest Quarter","Precipitation of Warmest Quarter","Precipitation of Coldest Quarter")
### create liner-regression list
plot_liner <- list()
for (i in 1:length(files_txt)){
  a=unlist(strsplit(files_txt[i],split="_"))
  gene_function=a[2]
  CHR=a[3]
  pos=a[4]
  type=a[5]
  GENE=a[6]
  REF=a[7]
  ALT=a[8]
  freq=read.table(files_txt[i],head=F)  
  temp=gsub("BIO","",a[9])
  env=gsub(".txt","",temp)
  BIO=paste("BIO",temp,sep="")
  n=as.numeric(env)
  selected_sites=data.frame(matrix(nrow=24))
  selected_sites$SN=current_data$SN
  selected_sites$current_env=current_data[,paste("current_BIO",env,sep="")]
  selected_sites$types=current_data$types
  selected_sites$Covar=current_data$Covar
  current_min=min(current_data[,paste("current_BIO",env,sep="")])
  current_max=max(current_data[,paste("current_BIO",env,sep="")])
  xmin=current_min
  xmax=current_max
  ### Ensure that the slope of the regression line is positive
  selected_sites$alt_freq=freq[,2]
  freq_max_current=subset(selected_sites,current_env==current_max)
  freq_min_current=subset(selected_sites,current_env==current_min)
  BASE="ALT"
  BASE_type=ALT
  if (freq_max_current$alt_freq < freq_min_current$alt_freq){
    selected_sites$alt_freq=1-selected_sites$alt_freq
    BASE="REF"
    BASE_type=REF
  }
  title_y=paste(BASE,"-frequency of ",CHR,"_",pos,"(",type,")",sep="")
  out_PDF=paste("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/freq/picture2/",gsub(".txt",".pdf",files_txt[i]),sep="")
  plot_liner[[n]] = ggplot(selected_sites,aes(current_env,alt_freq,color=factor(SN)))+
    geom_point(size=2.5)+
    scale_color_manual(values=c("#0a6690","#ff3333"))+
    geom_smooth(formula = y ~ x, method = "lm",se=F,level=0.95,color="black",size=0.7,fullrange = F)+
    geom_ribbon(stat = "smooth",
                method = "lm",
                se = TRUE,
                size=0.5,
                alpha = 0, # or, use fill = NA
                colour = "black",
                linetype = "dashed")+
    xlab(label = paste("Environment (","BIO",env,": ",text_data[n],")",sep=""))+
    ylab(label = title_y)+
    xlim(xmin,xmax)+
    scale_y_continuous(limits = c(0,1),breaks=seq(0,1,0.25))+
    #geom_hline(aes(yintercept=1), colour="#A0A0A0", linetype="twodash",size=0.8)+
    theme_bw()+
    theme(
      text=element_text(family="serif"),
      axis.ticks.length = unit(0.25,"lines"),axis.ticks=element_line(colour="black",unit(0.6,"line")),
      axis.text.x=element_text(angle=0,hjust=1,vjust=0.5,size=12,family="serif",colour = "black"),
      axis.text.y=element_text(size=12,family="serif",colour = "black"), 
      plot.title = element_text(
        size = 12L,
        hjust = 0.5
      ),
      axis.title.y = element_text(size = 12,family="serif",colour="black"),
      legend.key.size=unit(0.3,'inch'),
      legend.title = element_text(size=15, color="black",vjust=0.5, hjust=0.5,family="serif"),
      legend.position = "none",
      legend.text= element_text(size=10, color="black",vjust=0.5, hjust=0.5),
      axis.title.x = element_text(size = 13,family="serif"),
      panel.background=element_rect(fill="white"),
      plot.background = element_rect(fill = "white"),
      panel.grid.major =element_blank(), panel.grid.minor = element_blank(),
      plot.margin=unit(c(0.5,0.5,0.5,0.5),"mm"))
  ggsave(plot_liner[[n]],filename=out_PDF,width=5,height=5,dpi=600)
}

###########  gene_structure ###########
plot_gene <- list()
files_gff=list.files(path="F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/gene",full.names = F,pattern = "gff")
setwd("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/gene")
gene_list=read.csv("gene_list.csv",head=T)
for (i in c(1:4,6:12,14:19)){
  gene_symbol=subset(gene_list,BIO==i) 
  gene=gene_symbol$gene_name
  LG=gene_symbol$CHR
  gff_data=read.table(paste(gene,".gff",sep=""),head=F)
  colnames(gff_data)=c("n1","n2","n3","n4","n5","n6","n7","n8","n9")
  gff_data$gene=gene
  variation_start=gene_symbol$start/1000
  variation_end=gene_symbol$end/1000
  ### delete gene and mRNA
  conserve=c("CDS","five_prime_UTR","three_prime_UTR")
  gff_data=subset(gff_data[,c("n1","n3","n4","n5","n7","gene")],n3%in%conserve)
  gff_data$n1=LG
  colnames(gff_data)=c("molecule","type","start","end","strand","gene")
  gff_data$start=gff_data$start/1000
  gff_data$end=gff_data$end/1000
  if (gff_data[1,5]=="+"){
    gff_data$direction=1
  }else {
    gff_data$direction=-1
  } 
  type_color=c("#8dd3c7","#ffffb3","#bebada")
  names(type_color)=c("CDS","five_prime_UTR","three_prime_UTR")
  plot_gene[[i]]  = ggplot(gff_data, aes(xmin = start, xmax = end,y = gene, fill = type,label=type,forward = direction)) +
    #scale_fill_brewer(palette = "Set3") +
    scale_fill_manual(values=type_color[unique(gff_data$type)])+
    geom_vline(aes(xintercept=variation_start),color="red", linetype="dashed",size=0.8)+
    geom_vline(aes(xintercept=variation_end),color="red", linetype="dashed",size=0.8)+
    ylab(label = "")+xlab(label=paste(LG," (Kb)"))+
    geom_gene_arrow() +
    theme_genes()+
    geom_gene_label(label="",align = "left")+
    theme(text=element_text(family="serif"),
          axis.ticks.length = unit(0.25,"lines"),axis.ticks=element_line(colour="black",unit(0.6,"line")),
          axis.text.x=element_text(angle=0,hjust=1,vjust=0.5,size=11,family="serif",colour = "black"),
          axis.text.y=element_text(size=13,colour = "black"), 
          axis.title.y = element_text(size = 10,family="serif"),
          legend.key.size=unit(0.2,'inch'),
          legend.title = element_blank(),
          legend.position = "right",
          legend.text= element_text(size=10.2, color="black",vjust=0.5, hjust=0.5,family="serif"),
          axis.title.x = element_text(size = 12,family="serif"),
          panel.background=element_rect(fill="white"),
          plot.background = element_rect(fill = "white"),
          panel.grid.major =element_blank(), panel.grid.minor = element_blank(),
          plot.margin=unit(c(0.5,0.5,1,1),"mm"))
  ggsave(plot_gene[[i]],filename=paste("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/gene/picture3/",gene,".pdf",sep=""),width=10,height=1)
  ########### 4 plots merge ########### 
  pdf(paste("F:/360MoveData/Users/Desktop/pk_paper/picture/22gene_supp/merge_picture/BIO",i,".pdf",sep=""),width=10.5,height=7.5)
  grid.arrange(plot_manhatton[[i]],plot_gene[[i]],
               arrangeGrob(plot_map[[i]],plot_liner[[i]],ncol=2), nrow=3,heights=c(1.55/5,0.55/5,2.9/5))
  dev.off()
}

















